### DISCUSSION TITLE
Generating a Proof for a Private Network

### DISCUSSION THREAD

**Discussion by zd819** [zkonduit/ezkl#195] [Q&A] [answered]
Hey All. Hope you are doing well.

Running the default settings for the ezkl compiler work fine(public network, private input).
However, I am unsure about generating proofs for private networks, that is : proving for a private network that the public inputs into the network produce a specific output.
According to the documentation I believe this is possible: "I ran my private neural network on some public data and it produced this output", and would appreciate any help in doing this.

Thanks

**Comment by alexander-camuto** [comment]
Hey @zd819 :) by default the parameters of the network are private, the inputs to the network are private and the outputs of the network are publically known. 

You can play around with these settings with the following flags 
```bash
# sets the networks parameters to be public or private 
--public-params=true/false
# sets the network inputs to be public or private 
--public-inputs=true/false
# sets the network output to be public or private 
--public-outputs=true/false
```

Example usage for public inputs to the network produce a specific public output, whilst keeping weights private 

```bash

ezkl --public-inputs=true --public-params=false --public-outputs=true  prove ...

```

Hope that helps :)

**Reply by zd819** [reply]
Hi @alexander-camuto and thanks for the response.

I assumed the default was a public model, since in the Python and CLI Section, verification required the model file: 

`ezkl --bits=16 -K=17 prove -D ./examples/onnx/1l_relu/input.json -M ./examples/onnx/1l_relu/network.onnx --proof-path 1l_relu.pf --vk-path 1l_relu.vk --params-path=kzg.params`

I then tried setting --public-params=false : 
```
error: unexpected value 'false' for '--public-params' found; no more were expected

Usage: ezkl <--tolerance <TOLERANCE>|--scale <SCALE>|--bits <BITS>|--logrows <LOGROWS>|--public-inputs|--public-outputs <PUBLIC_OUTPUTS>|--public-params|--pack-base <PACK_BASE>|--check-mode <CHECK_MODE>> <COMMAND>
```

And when removing the --public-params flag altogether: 

`ezkl --bits=16 -K=17 verify --proof-path proof.bin --vk-path vk.json --params-path "../../../../ZK-ML/params/kzg.params`
```
error: the following required arguments were not provided:
  --model <MODEL>
```

Unless I am misinterpreting something the verification here requires the model (file), which defeats the idea of a private model since the verifier has access to it?
Please correct me if my conclusion here is wrong.
Thanks

**Reply by jasonmorton** [reply]
What you really need is a verification key, although it is not easy to see how to verify without the model itself. We do currently have support for outputting an evm verifier, which can be used to verify without access to the model. Working on other options for easily verifying without model access.

**Reply by alexander-camuto** [reply]
Would note that this is due to the way the Halo2 API is structured -- and we're working hard of getting a verification flow in rust that doesn't require access to the model file. 

A (hack) for now is to create a model file that has the same layers as the original model used in prove but with the parameters altered -- this will still allow you to verify on rust but will keep the params private. We really just need the "shape" info contained in the model to verify really -- not the parameters themselves. 

If you set public_params to false, those parameters will be hidden in the EVMVerifier. I also have made the flags explicitly settable as I described above in this PR https://github.com/zkonduit/ezkl/pull/201 . I think the toggling we had before was unintuitive. Let me know if that branch fixes: 

`error: unexpected value 'false' for '--public-params' found; no more were expected`

**Reply by jasonmorton** [reply]
I'd note that in fact in most cases the model architecture is largely hidden when you use the evm verifier.

**Reply by alexander-camuto** [reply]
Yeah -- you can also verify any proofs locally (without deploying the contract) using the evm verifier as well. 

There are detailed instructions on how to do so in the readme. Let me know if anything is unclear there.

**Reply by zd819** [reply]
Thanks for the replies and comments everyone. It makes sense now why the model is required for current verification, and also the evm verifier is very useful too. 
How far is the progress in allowing the compiler to verify without model access, and what are the main bottlenecks with getting it to work with the Halo2 APIs?
This use case especially is something I am very interested in and would look into it or help if possible.
Thanks again

**Reply by alexander-camuto** [reply]
We're getting closer every day aha. A big step towards this was in https://github.com/zkonduit/ezkl/pull/208 . Basically recent updates to Halo2 have allowed more things to be saved / serialized between runs. So hopefully we can get a verifier that doesn't require access to a model. If you are in need of this feature we can hop on a call to discuss it / open an issue to track it :)

**Reply by alexander-camuto** [reply, marked as answer]
Hey @zd819 I've implemented it here https://github.com/zkonduit/ezkl/pull/216 . Hope that helps

**Reply by zd819** [reply]
Hey @alexander-camuto thanks for the replies and the recent push. 
I would love to hop on a call if you have time to discuss the implementation as I have a couple of questions still. 
Whichever platform you prefer and whatever time you prefer I'll be available, just let me know.
Thanks

**Reply by alexander-camuto** [reply]
FInd my twitter and DM me (should be in my github bio) :)

**Reply by zd819** [reply]
Just followed, I think your DMs are turned off as I am unable to DM right now. My tag is @BlasianHokage
